hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECBlockOutputStreamEntry.java

```diff
@@ -17,53 +17,259 @@
+    Pipeline ecPipeline = getPipeline();
+    List<DatanodeDetails> nodes = getPipeline().getNodes();
+    blockOutputStreams =
+        new ECBlockOutputStream[nodes.size()];
+    for (int i = 0; i< getPipeline().getNodes().size(); i++) {
+      blockOutputStreams[i] = new ECBlockOutputStream(
+          getBlockID(),
+          getXceiverClientManager(),
+          createSingleECBlockPipeline(ecPipeline, nodes.get(i), i+1),
+          getBufferPool(),
+          getConf(),
+          getToken());
+    }
+  }
+
+  @Override
+  public OutputStream getOutputStream() {
+    if (!isInitialized()) {
+      return null;
+    }
+    return blockOutputStreams[currentStreamIdx];
+  }
+
+  @Override
+  boolean isInitialized() {
+    return blockOutputStreams != null;
+  }
+
+  public int getCurrentStreamIdx() {
+    return currentStreamIdx;
+  }
+
+  public void useNextBlockStream() {
+    currentStreamIdx++;
+  }
+
+  public void forceToFirstParityBlock(){
+    currentStreamIdx = replicationConfig.getData();
+  }
+
+  @Override
+  void incCurrentPosition() {
+    if (isWritingParity()) {
+      return;
+    }
+    super.incCurrentPosition();
+  }
+
+  @Override
+  void incCurrentPosition(long len) {
+    if (isWritingParity()){
+      return;
+    }
+    super.incCurrentPosition(len);
+  }
+
+  @Override
+  public void flush() throws IOException {
+    if (!isInitialized()) {
+      return;
+    }
+    for(int i=0; i<=currentStreamIdx && i<blockOutputStreams.length; i++) {
+      blockOutputStreams[i].flush();
+    }
+  }
+
+  @Override
+  boolean isClosed() {
+    if (!isInitialized()) {
+      return false;
+    }
+    return blockStreams().allMatch(BlockOutputStream::isClosed);
   }
 
   @Override
-  ECBlockOutputStream createOutputStream() throws IOException {
-    this.out = new ECBlockOutputStream(getBlockID(), getXceiverClientManager(),
-        getPipeline(), getBufferPool(), getConf(), getToken());
-    return this.out;
+  public void close() throws IOException {
+    if (!isInitialized()) {
+      return;
+    }
+    for (ECBlockOutputStream stream : blockOutputStreams) {
+      stream.close();
+    }
+    updateBlockID(underlyingBlockID());
+  }
+
+  @Override
+  long getTotalAckDataLength() {
+    if (!isInitialized()) {
+      return 0;
+    }
+    // blockID is the same for EC blocks inside one block group managed by
+    // this entry.
+    updateBlockID(underlyingBlockID());
+    //TODO: A future implementation might require something like this, but
+    // currently as ECBlockOutputStream is inheriting from BlockOutputStream
+    // this method returns 0 all the time from the unrelying streams.
+    // After we have a confirmed ack mechanism, like there is in
+    // RatisBlockOutputStream, we should revisit this part, and decide if we
+    // want to filter out parity here for example.
+//    return blockStreams()
+//        .mapToLong(BlockOutputStream::getTotalAckDataLength)
+//        .sum();
```

Comment:
Why are we returning 0?

Reply:
ECBlockOutputStream extends and extended BlockOutputStream before. Earlier ECBlockOutputStream did not override this method, so the method in BlockOutputStream was called. That method returned 0 hardwired.
RatisBlockOutputStream on the other hand maintains the acked data length via CommitWatcher, and returns that.

Our EC logic does not hold or check for any ack from the DNs via the grpc client, and does not maintain the acked legth so far, and it seems to be safer for now to just return 0 instead of calling the method of underlying streams, as we may not do this via those streams later on. I have added the naive implementation for discussion. I am unsure if the acked length should account for the parity block acks or for just the data block acks either.
And in any case, even if we sum the results, we will get 0 anyways, but it is easier to spot this way that the logic is not yet developed to calculate the acked length. Did I miss something?